yellow-paper/docs/contract-deployment/instances.md

```diff
@@ -17,102 +17,185 @@ The structure of a contract instance is defined as:
+
+Removing this check greatly simplifies the `ContractInstanceDeployer`, since it does not need to invoke constructors or check initialization nullifiers emitted from other contracts.
+
+```diff
+function deploy (
+  user_salt: Field, 
+  contract_class_id: Field, 
+  initialization_hash: Field, 
+  portal_contract_address: Field, 
+  public_keys_hash: Field, 
+  universal_deploy?: boolean,
+-  constructor_selector?: Field,
+-  constructor_args?: Field[],
+)
+  assert nullifier_exists silo(contract_class_id, ClassRegisterer)
+  assert is_valid_eth_address(portal_contract_address)
+  
+  deployer = if universal_deploy then this.address else msg_sender
+  salt = pedersen([user_salt, this.address], GENERATOR__ADDRESS_SALT)
+  version = 1
+  address = compute_address(version, salt, contract_class_id, initialization_hash, portal_contract_address, public_keys_hash)
+
+  emit_nullifier(address)
+-  if constructor_selector then call(address, constructor_selector, constructor_args)
+-  assert nullifier_exists silo(address, address)
+
+  emit_event ContractInstanceDeployed(address, version, salt, contract_class_id, initialization_hash, portal_contract_address, public_keys_hash)
+```
+
+This approach also allows devs to handle initialization in alternate ways, since there is nothing in the protocol or in a canonical contract that mandates how contracts are expected to be initialized. And it also allows for contracts to _not_ have a constructor at all if it is not needed, while still allowing them to be Publicly Deployed.
+
+However, this opens the door for a contract to be simultaneously Publicly Deployed and Uninitialized, which is a state that does not seem to map to a valid use case. Furthermore, it requires public functions to check the Initialization Nullifier on every call, which in the current approach is not needed as the presence of the Deployment Nullifier checked by the Public Kernel is enough of a guarantee that the contract was initialized.
+
+<!-- To @iAmMichaelConnor: I am tempted to actually implement this approach in spite of its drawbacks. It makes responsibilities so much more clearly defined: Deployer deploys, Initializer initializes, and there is no link between the two. It also allows removing constructors when they are not needed. And the added cost of checking the initialization nullifier from public functions doesn't seem too terrible. -->
+
+### Execute Initialization during Public Deployment only
+
+While it is appealing to allow a user to privately create a new contract instance and not reveal it to the world, we have not yet validated this use case. We could simplify deployment by relying on a single nullifier to track Initialization, and couple it with Public Deployment. Private functions can check initialization via the Deployment Nullifier emitted by the `ContractInstanceDeployer`.
```

Comment:
I think this is a valid use case.
Think of every contract that's agreed between any two businesses in the traditional world: the terms of that contract (the bytecode) must stay private.